src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -2080,4 +2082,44 @@ protected void notifyShutdown() {
         // No-op, handled in the suspendAll terminator
     }
 
+    // TODO: Core could use a QuietDownListener class with `onQuietDown` and `onCancelQuietDown` methods.
+    @Restricted(NoExternalUse.class)
+    @Extension
+    public static class QuietDownListener extends PeriodicWork {
+        private boolean wasQuietDown = Jenkins.get().isQuietingDown();
+
+        @Override
+        public long getRecurrencePeriod() {
+            return Main.isUnitTest
+                    ? TimeUnit.SECONDS.toMillis(1)
+                    : TimeUnit.MINUTES.toMillis(1);
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            boolean isQuietDown = Jenkins.get().isQuietingDown();
+            // Quieting down is handled by `CpsThreadGroup.scheduleRun`, so we just need to handle quiet mode being cancelled.
+            if (!isQuietDown && wasQuietDown) {
```

Comment:
`AtomicBoolean` can express this kind of phase transition a bit better, though in this case only one thread should ever be in this listener at once due to the way `PeriodicWork` works so it does not really matter.